### PR TITLE
Add a fallback to git diff command to get diff of a GitHub pull request 

### PR DESCRIPTION
**Pull request by massongit**
- [x] Updated Unreleased section in [CHANGELOG](https://github.com/reviewdog/reviewdog/blob/master/CHANGELOG.md) or it's not notable changes.

Fix https://github.com/reviewdog/reviewdog/issues/1696

Based on https://github.com/reviewdog/reviewdog/issues/1696#issuecomment-2028410272, I add a fallback to `git diff` command to get diff of a GitHub pull request.

### CODE REVIEW

**Comment by atiqueakhtar** [comment]
Hi @massongit ,

Thank you for investigating the fix for this issue. Are we on track to resolve the check failure `(reviewdog / reviewdog (github-check) (pull_request))` soon? We're eager to know if this can be addressed soon, as the resolution is critical for our pending PRs for which reviewdog is failing.

Appreciate your efforts on this.

**Comment by massongit** [comment]
I think it will take several days to weeks to resolve this issue.

**Comment by radiantshaw** [comment]
> Therefore, I leave it as is.

@massongit Does that mean this PR will not be merged anytime soon?

**Comment by massongit** [comment]
I think that it will be merged soon if the review passes.

**Comment by massongit** [comment]
@shogo82148 @haya14busa Please review it.

File: doghouse/server/github_checker.go [hunk 1]
```diff
@@ -20,9 +22,17 @@ type checkerGitHubClient struct {
...
 func (c *checkerGitHubClient) GetPullRequestDiff(ctx context.Context, owner, repo string, number int) ([]byte, error) {
-	opt := github.RawOptions{Type: github.Diff}
-	d, _, err := c.PullRequests.GetRaw(ctx, owner, repo, number, opt)
-	return []byte(d), err
+	pr, _, err := c.PullRequests.Get(ctx, owner, repo, number)
+	if err != nil {
+		return nil, err
+	}
+
+	bytes, err := exec.Command("git", "diff", "--find-renames", pr.GetBase().GetSHA(), pr.GetHead().GetSHA()).CombinedOutput()
```
Comment (haya14busa) [inline doghouse/server/github_checker.go:30]: Dogserver doesn't have git repository and it doesn't work.
Please revert the change.

File: service/github/github.go [hunk 2]
```diff
@@ -300,12 +301,17 @@ func (g *PullRequest) setPostedComment(ctx context.Context) error {
...
 func (g *PullRequest) Diff(ctx context.Context) ([]byte, error) {
-	opt := github.RawOptions{Type: github.Diff}
-	d, _, err := g.cli.PullRequests.GetRaw(ctx, g.owner, g.repo, g.pr, opt)
+	pr, _, err := g.cli.PullRequests.Get(ctx, g.owner, g.repo, g.pr)
 	if err != nil {
 		return nil, err
 	}
-	return []byte(d), nil
+
+	bytes, err := exec.Command("git", "diff", "--find-renames", pr.GetBase().GetSHA(), pr.GetHead().GetSHA()).CombinedOutput()
```
Comment (haya14busa) [inline service/github/github.go:309]: We need to compare to merge base of SHA to get the same diff as Pull request.

Example: https://github.com/reviewdog/reviewdog/blob/dd1167b11c4d5eee154151ac1626afe24fd7cb95/service/gitlab/gitlab_mr_diff.go#L63-L65

File: service/github/github.go [hunk 3]
```diff
@@ -300,12 +301,17 @@ func (g *PullRequest) setPostedComment(ctx context.Context) error {
 
 // Diff returns a diff of PullRequest.
 func (g *PullRequest) Diff(ctx context.Context) ([]byte, error) {
-	opt := github.RawOptions{Type: github.Diff}
-	d, _, err := g.cli.PullRequests.GetRaw(ctx, g.owner, g.repo, g.pr, opt)
```
Comment (haya14busa) [inline service/github/github.go:304]: Can we keep this code as-is and use git-diff solution only when API request failed for backward compatibility?

File: .github/workflows/reviewdog.yml [hunk 4]
```diff
@@ -15,6 +15,8 @@ jobs:
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0 # In order to take diff, commits other than the target are also required
```
Comment (haya14busa) [inline .github/workflows/reviewdog.yml:19]: Can we remove these changes and use GitHub API by default? (See the other comment)

**Review by haya14busa — CHANGES_REQUESTED** [review]
Thank you for your contribution.

It's a bit tough problem and I'm not sure what's the best way to fix it.

**Comment by haya14busa** [comment]
> CI reviewdog / reviewdog (github-check) (pull_request) seems to fail if a pull request is created from a forked repository.
Therefore, I leave it as is.

btw, it's actually unexpected. reviewdog should not fail from a forked repository too. 
I guess there are some bug about exit code handling 🤔

**Comment by massongit** [comment]
https://github.com/reviewdog/reviewdog/pull/1697#issuecomment-2030085520

I submit https://github.com/reviewdog/reviewdog/pull/1700 that fixes that issue.
However, accoding to https://github.com/reviewdog/reviewdog/pull/1697#discussion_r1555827335, it doesn't seem to be a desirable solution...

File: doghouse/server/github_checker.go (hunk 1 above)
Comment (massongit) [reply doghouse/server/github_checker.go]: I revert it: https://github.com/reviewdog/reviewdog/pull/1697/commits/d72ab6bf3df88c3244619648365d7114a7a1af3d

File: .github/workflows/reviewdog.yml (hunk 4 above)
Comment (massongit) [reply .github/workflows/reviewdog.yml]: I revert it: https://github.com/reviewdog/reviewdog/compare/d72ab6bf3df88c3244619648365d7114a7a1af3d...0fc18a2d37e4ccb5397c924bfdb26835f6270af2

File: service/github/github.go (hunk 2 above)
Comment (massongit) [reply service/github/github.go]: I apply it: https://github.com/reviewdog/reviewdog/pull/1697/commits/3c496f50759521c76f0d43f38d0eefaebf8819e9

File: service/github/github.go (hunk 3 above)
Comment (massongit) [reply service/github/github.go:304]: I apply it: https://github.com/reviewdog/reviewdog/compare/3c496f50759521c76f0d43f38d0eefaebf8819e9...e67af34299d57ea9b857d71cfccdcea2cf0cb5dc

**Comment by massongit** [comment]
https://github.com/reviewdog/reviewdog/pull/1697#issuecomment-2042785381
>btw, it's actually unexpected. reviewdog should not fail from a forked repository too.
>I guess there are some bug about exit code handling 🤔

I fixed it: https://github.com/reviewdog/reviewdog/pull/1707
Therefore, please also review it.

**Comment by atiqueakhtar** [comment]
Hi @massongit ,

Thank you for the quick updates in response to the feedback!

Could the reviewers @haya14busa and @shogo82148 please take a look at the recent changes when you have a moment?

Thank you for your efforts on this!

**Review by haya14busa — CHANGES_REQUESTED** [review]

File: service/github/github.go [hunk 5]
```diff
@@ -302,10 +303,30 @@ func (g *PullRequest) setPostedComment(ctx context.Context) error {
 func (g *PullRequest) Diff(ctx context.Context) ([]byte, error) {
 	opt := github.RawOptions{Type: github.Diff}
 	d, _, err := g.cli.PullRequests.GetRaw(ctx, g.owner, g.repo, g.pr, opt)
+	if err == nil {
+		return []byte(d), nil
+	}
+
+	pr, _, err := g.cli.PullRequests.Get(ctx, g.owner, g.repo, g.pr)
 	if err != nil {
 		return nil, err
```
[the remark this answers is not in the document]
Comment (massongit) [reply service/github/github.go:312]: https://github.com/reviewdog/reviewdog/blob/7644825553f0a1b7681448dce36f8dc7ac16d686/service/github/github.go#L305-L308
Is this a review of the above code?

[the remark this answers is not in the document]
Comment (massongit) [reply service/github/github.go:312]: I think that this line of code is a part of the git command flow.

[the remark this answers is not in the document]
Comment (haya14busa) [reply service/github/github.go:312]: Oh my bad, I misread it. 
Can you address the second point?

[the remark this answers is not in the document]
Comment (massongit) [reply service/github/github.go:312]: >I think that this line is a part of the git command flow.

https://github.com/reviewdog/reviewdog/blob/86c205edb68bba175f6dca979d71fd2d772c4126/service/github/github.go#L304-L308
This is the code of GitHub API flow.

https://github.com/reviewdog/reviewdog/blob/86c205edb68bba175f6dca979d71fd2d772c4126/service/github/github.go#L310-L329
This is the code of git command flow.

I wanted to say that you reviewed about the GitHub API flow while you pointed to the code in the git command flow.

[the remark this answers is not in the document]
Comment (massongit) [reply service/github/github.go:316]: https://github.com/reviewdog/reviewdog/blob/b47e57c7e28eab11441fa3f506827f8c7f77d15e/service/github/github.go#L305

I fixed the above code to fallback on 406 error and return err otherwise: https://github.com/reviewdog/reviewdog/pull/1697/commits/b47e57c7e28eab11441fa3f506827f8c7f77d15e

**Review by haya14busa — APPROVED** [review]
Thank you for your contribution :+1:
